pkg/util/rollback/rollback.go

```diff
@@ -0,0 +1,144 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package rollback
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+
+	"kubevirt.io/client-go/kubecli"
+)
+
+const (
+	rolledBackFromDeploymentIDAnnotation = "kubevirt.io/rolledback-from-deployment-id"
+)
+
+type DaemonSetRollbacker struct {
+	clientset kubecli.KubevirtClient
+}
+
+type historiesByRevision []*appsv1.ControllerRevision
+
+func (h historiesByRevision) Len() int      { return len(h) }
+func (h historiesByRevision) Swap(i, j int) { h[i], h[j] = h[j], h[i] }
+func (h historiesByRevision) Less(i, j int) bool {
+	return h[i].Revision < h[j].Revision
+}
+
+func NewDaemonSetRollbacker(clientset kubecli.KubevirtClient) *DaemonSetRollbacker {
+	return &DaemonSetRollbacker{clientset}
+}
+
+// controlledHistories returns all ControllerRevisions in namespace that selected by selector and owned by accessor
+func (r *DaemonSetRollbacker) controlledHistory(
+	ds *appsv1.DaemonSet,
+	selector labels.Selector) ([]*appsv1.ControllerRevision, error) {
+	var result []*appsv1.ControllerRevision
+	historyList, err := r.clientset.AppsV1().ControllerRevisions(ds.Namespace).List(
+		context.Background(),
+		metav1.ListOptions{LabelSelector: selector.String()},
+	)
+	if err != nil {
+		return nil, err
+	}
+	for i := range historyList.Items {
+		history := historyList.Items[i]
+		if metav1.IsControlledBy(&history, ds) {
+			result = append(result, &history)
+		}
+	}
+	return result, nil
+}
+
+func (r *DaemonSetRollbacker) daemonSetHistory(ds *appsv1.DaemonSet) ([]*appsv1.ControllerRevision, error) {
+	selector, err := metav1.LabelSelectorAsSelector(ds.Spec.Selector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create selector for DaemonSet %s: %v", ds.Name, err)
+	}
+	history, err := r.controlledHistory(ds, selector)
+	if err != nil {
+		return nil, fmt.Errorf("unable to find history controlled by DaemonSet %s: %v", ds.Name, err)
+	}
+	return history, nil
+}
+
+func HasDaemonSetRolledBackFromDeployment(daemonSet *appsv1.DaemonSet, fromDeploymentID string) bool {
+	return daemonSet.Annotations[rolledBackFromDeploymentIDAnnotation] == fromDeploymentID
+}
+
+func (r *DaemonSetRollbacker) Rollback(daemonSet *appsv1.DaemonSet, sourceDeploymentID string) error {
+	history, err := r.daemonSetHistory(daemonSet)
+	if err != nil {
```

Comment:
in the event of a canary failure, I don't think the virt-operator should take any automated action to roll anything back. The admin has "declared" the version of KubeVirt they want. The virt-operator should block (and report) when it can't converge on the desired state. It's then the admin's choice to determine how to proceed.

With the canary pod we're simply attempting to reduce the blast radius, not make any decisions about how to correct the issue.

Reply:
So no rollback at all? I guess it makes sense to not take any automated action against what the user declared but it kinda sounded nice to have as it  would restore the node to a functioning state.

Reply:
If the rollback is not needed at all then everything becomes simpler, I probably misunderstood as I took it as part of the specs to solve the slow upgrade issue.

Reply:
yup, automated rollback isn't necessary. If the upgrade encounters an issue, the expectation is that the issue is reported (via condition, events, etc...) and that intervention is required to decide how to correct the issue. Someone could create their own automation to rollback kubevirt if update is blocked or takes too long, but our virt-operator will only try to continually converge on the declared intent within the KubeVirt CR.

Reply:
Right, got it, I'll modify and re-push, thank you for taking a look!